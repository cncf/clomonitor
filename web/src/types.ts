export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
