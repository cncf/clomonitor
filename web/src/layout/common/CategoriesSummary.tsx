import classNames from 'classnames';

import { CATEGORY_ICONS } from '../../data';
import { Score, ScoreType } from '../../types';
import styles from './CategoriesSummary.module.css';
import CategoryProgressbar from './CategoryProgressbar';
import RoundScore from './RoundScore';

interface Props {
  score: Score;
  bigSize: boolean;
}

const CategoriesSummary = (props: Props) => {
  return (
    <div
      className={classNames(
        'align-items-center d-flex flex-column flex-md-row',
        styles.summary,
        {
          'flex-lg-column flex-xl-row': !props.bigSize,
        },
        { [styles.bigSize]: props.bigSize }
      )}
    >
      <div
        className={classNames(
          'd-none d-md-block',
          { 'd-lg-none d-xl-block d-lg-none d-xl-block': !props.bigSize },
          { 'mx-3': props.bigSize }
        )}
      >
        <div className="d-flex flex-column me-0 me-sm-4 mb-2 mb-sm-0">
          <RoundScore score={props.score.global} />
        </div>
      </div>

      <div
        className={classNames('flex-grow-1 w-100 position-relative', {
          'px-0 px-sm-3': props.bigSize,
        })}
      >
        <div className={classNames('row', { 'gx-4 gx-md-5': props.bigSize })}>
          <CategoryProgressbar
            name="Documentation"
            value={props.score.documentation}
            icon={CATEGORY_ICONS[ScoreType.Documentation]}
            bigSize={props.bigSize}
          />
          <CategoryProgressbar
            name="License"
            value={props.score.license}
            icon={CATEGORY_ICONS[ScoreType.License]}
            bigSize={props.bigSize}
          />
          <CategoryProgressbar
            name="Best Practices"
            value={props.score.best_practices}
            icon={CATEGORY_ICONS[ScoreType.BestPractices]}
            bigSize={props.bigSize}
          />
          <CategoryProgressbar
            name="Security"
            value={props.score.security}
            icon={CATEGORY_ICONS[ScoreType.Security]}
            bigSize={props.bigSize}
          />
          <CategoryProgressbar
            name="Legal"
            value={props.score.legal}
            icon={CATEGORY_ICONS[ScoreType.Legal]}
            bigSize={props.bigSize}
          />
        </div>
      </div>
    </div>
  );
};

export default CategoriesSummary;
