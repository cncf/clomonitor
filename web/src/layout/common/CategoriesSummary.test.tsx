import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import { ScoreType } from '../../types';
import CategoriesSummary from './CategoriesSummary';

const defaultProps = {
  score: {
    [ScoreType.BestPractices]: 95,
    [ScoreType.Documentation]: 85,
    [ScoreType.Global]: 65,
    [ScoreType.License]: 80,
    [ScoreType.Security]: 0,
    [ScoreType.Legal]: 75,
  },
  bigSize: false,
};

describe('CategoriesSummary', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Router>
        <CategoriesSummary {...defaultProps} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(
        <Router>
          <CategoriesSummary {...defaultProps} />
        </Router>
      );

      expect(screen.getByText('Documentation')).toBeInTheDocument();
      expect(screen.getByText('License')).toBeInTheDocument();
      expect(screen.getByText('Best Practices')).toBeInTheDocument();
      expect(screen.getByText('Security')).toBeInTheDocument();
      expect(screen.getByText('Legal')).toBeInTheDocument();

      expect(screen.getByTestId('global-score')).toBeInTheDocument();
      expect(screen.getAllByTestId('line')).toHaveLength(5);
    });

    it('renders correct classes when bigSize is true', () => {
      const { container } = render(
        <Router>
          <CategoriesSummary {...defaultProps} bigSize />
        </Router>
      );

      expect(container.children[0]).toHaveClass('bigSize');
      expect(container.children[0].children[1]).toHaveClass('px-0 px-sm-3');
      expect(container.children[0].children[1].children[0]).toHaveClass('gx-4 gx-md-5');
      expect(screen.getByText('85')).toHaveClass('bigSize');
    });
  });
});
