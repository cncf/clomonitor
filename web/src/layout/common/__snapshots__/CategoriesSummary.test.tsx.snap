// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesSummary creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="align-items-center d-flex flex-column flex-md-row summary flex-lg-column flex-xl-row"
  >
    <div
      class="d-none d-md-block d-lg-none d-xl-block d-lg-none d-xl-block"
    >
      <div
        class="d-flex flex-column me-0 me-sm-4 mb-2 mb-sm-0"
      >
        <div
          class="d-flex align-items-center justify-content-center rounded-pill RoundScore-module_score__AbYPP undefined"
          data-testid="global-score"
        >
          <div
            class="RoundScore-module_chart__s05eS"
          >
            <svg
              class="d-block"
              viewBox="0 0 36 36"
            >
              <path
                class="RoundScore-module_circleBg__3qRLN"
                d="M18 2.0845
          a 15.9155 15.9155 0 0 1 0 31.831
          a 15.9155 15.9155 0 0 1 0 -31.831"
              />
              <path
                class="RoundScore-module_circle__C-6ak"
                d="M18 2.0845
          a 15.9155 15.9155 0 0 1 0 31.831
          a 15.9155 15.9155 0 0 1 0 -31.831"
                stroke-dasharray="65, 100"
                style="stroke: var(--clo-yellow);"
              />
              <text
                class="RoundScore-module_value__mx3Fe"
                x="18"
                y="23"
              >
                65
              </text>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex-grow-1 w-100 position-relative"
    >
      <div
        class="row"
      >
        <div
          class="wrapper col-12"
        >
          <div
            class="d-flex flex-row bg-white position-relative border border-1 overflow-hidden"
          >
            <div
              class="d-flex flex-row align-items-center text-muted fw-bold flex-nowrap px-1 my-auto title"
            >
              <span
                class="pe-1 d-inline-block position-relative icon"
              >
                <svg
                  aria-hidden="true"
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
              <span
                class="text-truncate"
              >
                Documentation
              </span>
            </div>
            <div
              class="text-center fw-bold font-monospace value "
            >
              85
            </div>
            <div
              class="flex-grow-1 position-relative mx-2 progressWrapper  "
            >
              <div
                class="position-absolute start-0 top-0 bottom-0 line"
                data-testid="line"
                style="width: 85%;"
              />
            </div>
          </div>
        </div>
        <div
          class="wrapper col-12"
        >
          <div
            class="d-flex flex-row bg-white position-relative border border-1 overflow-hidden"
          >
            <div
              class="d-flex flex-row align-items-center text-muted fw-bold flex-nowrap px-1 my-auto title"
            >
              <span
                class="pe-1 d-inline-block position-relative icon"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 32c-70.7 0-128 57.3-128 128s57.3 128 128 128 128-57.3 128-128S326.7 32 256 32zm0 208c-44.2 0-80-35.8-80-80s35.8-80 80-80 80 35.8 80 80-35.8 80-80 80zM193.7 307.4c-19.1-8.1-36.2-19.6-50.8-34.3-1.4-1.4-2.8-2.8-4.1-4.3L64 400h96l48 80 48-105.8 25.5-56.2c-8.4 1.3-16.9 2-25.5 2-21.6 0-42.5-4.2-62.3-12.6zM373.3 268.9c-1.3 1.4-2.7 2.9-4.1 4.3-14.6 14.6-31.7 26.2-50.7 34.2L294 361.2l-21.9 48.4L304 480l48-80h96l-74.7-131.1z"
                  />
                </svg>
              </span>
              <span
                class="text-truncate"
              >
                License
              </span>
            </div>
            <div
              class="text-center fw-bold font-monospace value "
            >
              80
            </div>
            <div
              class="flex-grow-1 position-relative mx-2 progressWrapper  "
            >
              <div
                class="position-absolute start-0 top-0 bottom-0 line"
                data-testid="line"
                style="width: 80%;"
              />
            </div>
          </div>
        </div>
        <div
          class="wrapper col-12"
        >
          <div
            class="d-flex flex-row bg-white position-relative border border-1 overflow-hidden"
          >
            <div
              class="d-flex flex-row align-items-center text-muted fw-bold flex-nowrap px-1 my-auto title"
            >
              <span
                class="pe-1 d-inline-block position-relative icon"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5 4.60434V13.7889C5 15.1263 5.6684 16.3752 6.7812 17.1171L12 20.5963L17.2188 17.1171C18.3316 16.3752 19 15.1263 19 13.7889V4.60434L12 3.04879L5 4.60434ZM3.78307 2.82598L12 1L20.2169 2.82598C20.6745 2.92766 21 3.33347 21 3.80217V13.7889C21 15.795 19.9974 17.6684 18.3282 18.7812L12 23L5.6718 18.7812C4.00261 17.6684 3 15.795 3 13.7889V3.80217C3 3.33347 3.32553 2.92766 3.78307 2.82598ZM12 13.5L9.06107 15.0451L9.62236 11.7725L7.24472 9.45492L10.5305 8.97746L12 6L13.4695 8.97746L16.7553 9.45492L14.3776 11.7725L14.9389 15.0451L12 13.5Z"
                  />
                </svg>
              </span>
              <span
                class="text-truncate"
              >
                Best Practices
              </span>
            </div>
            <div
              class="text-center fw-bold font-monospace value "
            >
              95
            </div>
            <div
              class="flex-grow-1 position-relative mx-2 progressWrapper  "
            >
              <div
                class="position-absolute start-0 top-0 bottom-0 line"
                data-testid="line"
                style="width: 95%;"
              />
            </div>
          </div>
        </div>
        <div
          class="wrapper col-12"
        >
          <div
            class="d-flex flex-row bg-white position-relative border border-1 overflow-hidden"
          >
            <div
              class="d-flex flex-row align-items-center text-muted fw-bold flex-nowrap px-1 my-auto title"
            >
              <span
                class="pe-1 d-inline-block position-relative icon"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C9.243 2 7 4.243 7 7v2H6c-1.103 0-2 .897-2 2v9c0 1.103.897 2 2 2h12c1.103 0 2-.897 2-2v-9c0-1.103-.897-2-2-2h-1V7c0-2.757-2.243-5-5-5zM9 7c0-1.654 1.346-3 3-3s3 1.346 3 3v2H9V7zm9.002 13H13v-2.278c.595-.347 1-.985 1-1.722 0-1.103-.897-2-2-2s-2 .897-2 2c0 .736.405 1.375 1 1.722V20H6v-9h12l.002 9z"
                  />
                </svg>
              </span>
              <span
                class="text-truncate"
              >
                Security
              </span>
            </div>
            <div
              class="text-center fw-bold font-monospace value "
            >
              0
            </div>
            <div
              class="flex-grow-1 position-relative mx-2 progressWrapper  "
            >
              <div
                class="position-absolute start-0 top-0 bottom-0 line"
                data-testid="line"
                style="width: 1%;"
              />
            </div>
          </div>
        </div>
        <div
          class="wrapper col-12"
        >
          <div
            class="d-flex flex-row bg-white position-relative border border-1 overflow-hidden"
          >
            <div
              class="d-flex flex-row align-items-center text-muted fw-bold flex-nowrap px-1 my-auto title"
            >
              <span
                class="pe-1 d-inline-block position-relative icon"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12.75 2.75V4.5h1.975c.351 0 .694.106.984.303l1.697 1.154c.041.028.09.043.14.043h4.102a.75.75 0 0 1 0 1.5H20.07l3.366 7.68a.749.749 0 0 1-.23.896c-.1.074-.203.143-.31.206a6.296 6.296 0 0 1-.79.399 7.349 7.349 0 0 1-2.856.569 7.343 7.343 0 0 1-2.855-.568 6.205 6.205 0 0 1-.79-.4 3.205 3.205 0 0 1-.307-.202l-.005-.004a.749.749 0 0 1-.23-.896l3.368-7.68h-.886c-.351 0-.694-.106-.984-.303l-1.697-1.154a.246.246 0 0 0-.14-.043H12.75v14.5h4.487a.75.75 0 0 1 0 1.5H6.763a.75.75 0 0 1 0-1.5h4.487V6H9.275a.249.249 0 0 0-.14.043L7.439 7.197c-.29.197-.633.303-.984.303h-.886l3.368 7.68a.75.75 0 0 1-.209.878c-.08.065-.16.126-.31.223a6.077 6.077 0 0 1-.792.433 6.924 6.924 0 0 1-2.876.62 6.913 6.913 0 0 1-2.876-.62 6.077 6.077 0 0 1-.792-.433 3.483 3.483 0 0 1-.309-.221.762.762 0 0 1-.21-.88L3.93 7.5H2.353a.75.75 0 0 1 0-1.5h4.102c.05 0 .099-.015.141-.043l1.695-1.154c.29-.198.634-.303.985-.303h1.974V2.75a.75.75 0 0 1 1.5 0ZM2.193 15.198a5.414 5.414 0 0 0 2.557.635 5.414 5.414 0 0 0 2.557-.635L4.75 9.368Zm14.51-.024c.082.04.174.083.275.126.53.223 1.305.45 2.272.45a5.847 5.847 0 0 0 2.547-.576L19.25 9.367Z"
                  />
                </svg>
              </span>
              <span
                class="text-truncate"
              >
                Legal
              </span>
            </div>
            <div
              class="text-center fw-bold font-monospace value "
            >
              75
            </div>
            <div
              class="flex-grow-1 position-relative mx-2 progressWrapper  "
            >
              <div
                class="position-absolute start-0 top-0 bottom-0 line"
                data-testid="line"
                style="width: 75%;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
